src/libraries/Microsoft.PowerFx.Core/Public/Values/RecordValue.cs

```diff
@@ -15,51 +15,135 @@ namespace Microsoft.PowerFx.Core.Public.Values
     /// </summary>
     public abstract class RecordValue : ValidFormulaValue
     {
-        public abstract IEnumerable<NamedValue> Fields { get; }
+        /// <summary>
+        /// Fields and their values directly available on this record. 
+        /// The field names should match the names on <see cref="Type"/>. 
+        /// </summary>
+        public IEnumerable<NamedValue> Fields
+        {
+            get
+            {
+                var fields = Type.GetNames();
+                foreach (var kv in fields)
+                {
+                    var fieldName = kv.Name;
+                    var fieldType = kv.Type;
+
+                    var value = GetField(fieldType, fieldName);
+                    yield return new NamedValue(fieldName, value);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RecordValue"/> class.
+        /// </summary>
+        /// <param name="type"></param>
+        public RecordValue(RecordType type) 
+            : base(IRContext.NotInSource(type))
+        {
+        }
 
         internal RecordValue(IRContext irContext)
             : base(irContext)
         {
             Contract.Assert(IRContext.ResultType is RecordType);
         }
 
-        // If we have a derived Value, we can get a derived type. 
+        /// <summary>
+        /// The RecordType of this value.
+        /// </summary>
         public new RecordType Type => (RecordType)base.Type;
 
         public static RecordValue Empty()
         {
             var type = new RecordType();
-            return new InMemoryRecordValue(IRContext.NotInSource(type), new List<NamedValue>());
+            return new InMemoryRecordValue(IRContext.NotInSource(type), new Dictionary<string, FormulaValue>());
         }
-         
+
         /// <summary>
-        /// Get a field on this record. 
-        /// See <see cref="ObjectMarshaller"/> to override GetField behavior. 
+        /// Get a field on this record.         
         /// </summary>
-        /// <param name="name"></param>
-        /// <returns></returns>
-        public FormulaValue GetField(string name)
-        {            
-            var fieldType = Type.MaybeGetFieldType(name) ?? FormulaType.Blank;
+        /// <param name="fieldName">Name of field on this record.</param>
+        /// <returns>Field value or blank if missing. </returns>
+        public FormulaValue GetField(string fieldName)
+        {
+            if (fieldName == null)
+            {
+                throw new ArgumentNullException(nameof(fieldName));
+            }
+
+            var fieldType = Type.MaybeGetFieldType(fieldName) ?? FormulaType.Blank;
```

Comment:
why would this ever fail?

Reply:
I did a test run and here are 2 cases where it fails:
>> Index([1, 2, 3, 4, 5], Error({Kind: 11}))

>> Filter(Table({a:1},{a:2},If(1<0,{a:3}),{a:4},{a:5}), IsBlank(a) Or a > 0)